### PR TITLE
Clean up log listener socket during shutdown

### PR DESCRIPTION
**Pull request by ReubenBond**
During shutdown, we should close the logging sockets and wait for them to close before continuing.
This avoids an `ObjectDisposedException` thrown while resolving loggers in the socket logging loop after the `IServiceProvider` has been disposed.

cc @danegsta
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2326)

### CODE REVIEW

File: src/Aspire.Hosting/Dcp/DcpHostService.cs [hunk 1]
```diff
@@ -197,23 +199,31 @@ private static Socket CreateLoggingSocket(string socketPath)
...
+
+        await Task.WhenAll(outputLoggers).ConfigureAwait(false);
+        socket.Dispose();
     }
 
-    private async Task LogSocketOutputAsync(Socket socket)
+    private async Task LogSocketOutputAsync(Socket socket, CancellationToken cancellationToken)
     {
-        var reader = PipeReader.Create(new NetworkStream(socket));
+        await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dcp/DcpHostService.cs:223]: Add a comment why this is here

Why is this better than calling method in a `Task.Run`?

File: src/Aspire.Hosting/Dcp/DcpHostService.cs [hunk 2]
```diff
@@ -70,9 +68,13 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
...
     }
 
-    public Task StopAsync(CancellationToken cancellationToken = default)
+    public async Task StopAsync(CancellationToken cancellationToken = default)
     {
-        return Task.CompletedTask;
+        _shutdownCts.Cancel();
+        if (_logProcessorTask is { } task)
+        {
+            await task.ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dcp/DcpHostService.cs:76]: Can this be changed to ignore expected errors (I'm guessing `OperationCanceledException`) and log unexpected errors?

File: src/Aspire.Hosting/Dcp/DcpHostService.cs [hunk 3]
```diff
@@ -197,23 +199,31 @@ private static Socket CreateLoggingSocket(string socketPath)
 
     private async Task StartLoggingSocketAsync(Socket socket)
     {
-        while (true)
+        await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dcp/DcpHostService.cs:202]: Add a comment why this is here

File: src/Aspire.Hosting/Dcp/DcpHostService.cs (hunk 2 above)
Comment (ReubenBond) [reply src/Aspire.Hosting/Dcp/DcpHostService.cs]: Sure thing, updated

File: src/Aspire.Hosting/Dcp/DcpHostService.cs (hunk 3 above)
Comment (ReubenBond) [reply src/Aspire.Hosting/Dcp/DcpHostService.cs]: Replaced with Task.Run instead

File: src/Aspire.Hosting/Dcp/DcpHostService.cs [hunk 4]
```diff
@@ -73,7 +73,17 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
...
+            try
+            {
+                await task.ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+            }
+            catch (Exception ex)
+            {
+                Console.Error.WriteLine($"Error in logging socket processor: {ex}");
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dcp/DcpHostService.cs:85]: ILogger would be better than console unless there is a reason for console here.

Comment (ReubenBond) [reply src/Aspire.Hosting/Dcp/DcpHostService.cs:85]: I am following the pattern used elsewhere in the file, but I see now that current usage is split between Console & ILogger

**Review by JamesNK — APPROVED** [review]

**Pull request merged**